specs/bep-v1.rst

```diff
@@ -196,331 +179,163 @@ For C=1:
-    /                                                               /
-    \                     Key (variable length)                     \
-    /                                                               /
-    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-    |                        Length of Value                        |
-    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-    /                                                               /
-    \                    Value (variable length)                    \
-    /                                                               /
-    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+All data within the message (post decompression, if compression is in use)
+MUST be in protocol buffer (version 3) encoding. All strings MUST use the
+Unicode UTF-8 encoding, normalization form C.
 
-Fields (ClusterConfigMessage)
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Cluster Config (Type = 0)
+^^^^^^^^^^^^^^^^^^^^^^^^^
 
-.. Documentation note: the first time a field is mentioned it is put in **bold
-   text**. We use the Space Separated names in running text and ASCII art
-   diagrams, and CamelCase in the XDR syntax block at the end.
+.. Documentation note: the structure of a message section is always:
+   1. A short description of the message
+   2. Protocol buffer schema of the message
+   3. Description of the fields in the message.
 
-The **Folders** field contains the list of folders that will be synchronized
-over the current connection.
+This informational message provides information about the cluster
+configuration as it pertains to the current connection. A Cluster Config
+message MUST be the first message sent on a BEP connection. Additional
+Cluster Config messages MUST NOT be sent after the initial exchange.
 
-The **Options** field is a list of options that apply to the current
-connection. The options are used in an implementation specific manner. The
-options list is conceptually a map of keys to values, although it is
-transmitted in the form of a list of key and value pairs, both of string type.
-Key ID:s are implementation specific. An implementation MUST ignore unknown
-keys. An implementation MAY impose limits on the length keys and values. The
-options list may be used to inform devices of relevant local configuration
-options such as rate limiting or make recommendations about request
-parallelism, device priorities, etc. An empty options list is valid for
-devices not having any such information to share. Devices MAY NOT make any
-assumptions about peers acting in a specific manner as a result of sent
-options.
-
-
-Fields (Folder Structure)
-~~~~~~~~~~~~~~~~~~~~~~~~~
+Protocol Buffer Schema
+~~~~~~~~~~~~~~~~~~~~~~
 
-The **ID** field contains the folder ID, as a human readable string.
+.. code-block:: proto
+
+    message ClusterConfigMessage {
+        repeated Folder folders = 1;
+
+        message Folder {
```

Comment:
Do we want to nest messages in the proto definition in the spec? As far as I can understand the encoding is identical and the nesting only causes type naming to differ. We don't want to use it in production as the generated type names are ugly (but https://github.com/gogo/protobuf/issues/186), but perhaps it makes the structure clearer in the spec? Or perhaps not?